deob-bytecode/src/main/kotlin/org/openrs2/deob/bytecode/transform/StringDecryptionTransformer.kt

```diff
@@ -0,0 +1,183 @@
+import org.objectweb.asm.tree.FieldInsnNode
+import org.objectweb.asm.tree.FieldNode
+import org.objectweb.asm.tree.LdcInsnNode
+import org.objectweb.asm.tree.MethodInsnNode
+import org.objectweb.asm.tree.TableSwitchInsnNode
+import org.openrs2.asm.classpath.ClassPath
+import org.openrs2.asm.deleteExpression
+import org.openrs2.asm.intConstant
+import org.openrs2.asm.transform.Transformer
+
+@Singleton
+public class StringDecryptionTransformer : Transformer() {
+    private var stringsDecrypted = 0
+    private var classesDecrypted = 0
+    private var stringsInlined = 0
+
+    override fun preTransform(classPath: ClassPath) {
+        for (library in classPath.libraries) {
+            for (clazz in library) {
+                val clinit = clazz.methods.find { it.name == "<clinit>" } ?: continue
+
+                // step 1: identify the decryption methods
+                val inner = clinit.instructions.iterator().asSequence().filterIsInstance<MethodInsnNode>()
+                    .filter {
+                        it.owner == clazz.name && it.desc == "(Ljava/lang/String;)[C" &&
+                            it.next?.opcode == Opcodes.INVOKESTATIC
+                    }.firstOrNull()?.let {
+                        clazz.methods.find { method -> method.name == it.name && method.desc == it.desc }
+                    }
+                val outer = clinit.instructions.iterator().asSequence().filterIsInstance<MethodInsnNode>()
+                    .filter {
+                        it.owner == clazz.name && it.desc == "([C)Ljava/lang/String;" &&
+                            it.previous?.opcode == Opcodes.INVOKESTATIC
+                    }.firstOrNull()?.let {
+                        clazz.methods.find { method -> method.name == it.name && method.desc == it.desc }
+                    }
+
+                if (inner == null || outer == null) continue
+
+                // step 2: get the xor table from the outer method
+                val switch: TableSwitchInsnNode = outer.instructions.iterator().asSequence()
+                    .filterIsInstance<TableSwitchInsnNode>().firstOrNull() ?: continue
+
+                val xorLookup = ArrayList<Number>()
+                for (label in switch.labels) {
+                    xorLookup.add(label.next.intConstant ?: continue)
+                }
+                xorLookup.add(switch.dflt.next.intConstant ?: continue)
+
+                // step 3: decrypt all strings in the initializer
+                val strings = ArrayList<String>()
+                var field: FieldNode? = null
+                var fieldInit: AbstractInsnNode? = null
+                for (insn in clinit.instructions) {
+                    if (insn.opcode != Opcodes.INVOKESTATIC) continue
+                    val invoke = insn as MethodInsnNode
+                    if (invoke.owner != clazz.name || invoke.name != outer.name || invoke.desc != outer.desc) continue
+
+                    val ldc = invoke.previous.previous as LdcInsnNode
```

Comment:
The use of stuff like `.previous.previous` is a hint that this should probably use InsnMatcher - I'd use something like `InsnMatcher.compile("LDC INVOKESTATIC INVOKESTATIC AASTORE? PUTSTATIC")`.

Reply:
though it might also be cleaner to separate the PUTSTATIC (which happens once) from the AASTOREs (which happen multiple times) - idk what the exact bytecode pattern looks like though.